### PR TITLE
Treeview: Add a new selection type to support the permission tree

### PR DESCRIPTION
**Pull request by wisamidris7**
This Is For Supporting A Different Selection Type

### CODE REVIEW

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 1]
```diff
@@ -281,7 +281,9 @@ public void UpdateSelected(TKey key, bool isSelected)
             nodeState.IsSelected = isSelected;
             nodeState.IsIndeterminate = false;
 
-            if (SelectionType == SelectionType.Leaf)
+            if (SelectionType == SelectionType.Standalone)
```
Comment (wisamidris7) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:284]: This To Support The Standalone For Permissions Tree

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 2]
```diff
@@ -292,9 +294,9 @@ public async Task EmitSelectedAsync()
         {
             var selected = Nodes.Values.Where(r =>
             {
-                if (SelectionType == SelectionType.Independent)
+                if (SelectionType == SelectionType.Independent || SelectionType == SelectionType.Standalone)
```
Comment (wisamidris7) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:297]: Here I Want To Make Standalone Same Independent But It Effect His Children But His Children Can't Effect Him

File: src/Component/BlazorComponent/Components/Treeview/SelectionType.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
     public enum SelectionType
     {
         Leaf,
-        Independent
+        Independent,
```
Comment (wisamidris7) [inline src/Component/BlazorComponent/Components/Treeview/SelectionType.cs:6]: I Added Enum Type

**Comment by capdiem** [comment]
@wisamidris7 Hey, let me guess your usage scenario: after selecting a child node, its ancestor nodes will be selected, just like in `SelectionType=SelectionType.Leaf`. However, the selected items returned by `SelectionType.Leaf` will not include ancestor nodes, and you want to include ancestor nodes, right?

image

**Comment by capdiem** [comment]
@wisamidris7 If so, I don't recommend adding a new SelectionType. Maybe add a new property of boolean type, which can return the ancestor node of the selected item when turned on.

**Comment by wisamidris7** [comment]
This Is Anathor Type Cause It's Like Leaf But If I Off All Children Children Won't Effect Their Parent But Parent Can Effect All He's Children

And Parent No Need To Have The Third The Minus Icon_

**Comment by wisamidris7** [comment]
But When I See That It's Not Gonna Be A Boolean Cause Selection Type Is Better

**Comment by wisamidris7** [comment]
Or Do It By Your Self But What I Care About Is The Same Result

**Comment by wisamidris7** [comment]
For Example Like This

https://github.com/masastack/MASA.Blazor/assets/104096256/890621cf-7948-4a0f-8fb2-d40a50cf9abc

**Comment by capdiem** [comment]
I see.

**Comment by wisamidris7** [comment]
> I see.

I don't understand what gonna happen

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs (hunk 2 above)
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:297]: I think `LeafButIndependentParent` is better than `Standalone`.

```suggestion
                if (SelectionType is SelectionType.Independent or SelectionType.LeafButIndependentParent)
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 4]
```diff
@@ -543,7 +545,7 @@ private void BuildTree(List<TItem> items, TKey? parent)
 
                 BuildTree(children, key);
 
-                if (SelectionType != SelectionType.Independent && parent != null && !Nodes.ContainsKey(key) &&
+                if (SelectionType != SelectionType.Independent && SelectionType != SelectionType.Standalone && parent != null && !Nodes.ContainsKey(key) &&
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:548]:
```suggestion
                if (SelectionType == SelectionType.Leaf && parent != null && !Nodes.ContainsKey(key) &&
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 5]
```diff
@@ -565,7 +567,7 @@ private void BuildTree(List<TItem> items, TKey? parent)
 
                 // TODO: there is still some logic in Vuetify but no implementation here, it's necessarily?
 
-                if (newNode.IsSelected && (SelectionType == SelectionType.Independent || !newNode.Children.Any()))
+                if (newNode.IsSelected && ((SelectionType == SelectionType.Independent || SelectionType == SelectionType.Standalone) || !newNode.Children.Any()))
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:570]:
```suggestion
                if (newNode.IsSelected && (SelectionType != SelectionType.Leaf || !newNode.Children.Any()))
```

File: src/Component/BlazorComponent/Components/Treeview/SelectionType.cs (hunk 3 above)
Comment (capdiem) [reply src/Component/BlazorComponent/Components/Treeview/SelectionType.cs:6]: `LeafButIndependentParent` is better

**Review by capdiem — CHANGES_REQUESTED** [review]
in `UpdateParentSelected()`:

```diff
- if (isIndeterminate)
+ if (SelectionType == SelectionType.LeafButIndependentParent)
+ {
+   nodeState.IsSelected = true;
+   nodeState.IsIndeterminate = false;
+ }
+ else if (isIndeterminate)
```

**Comment by capdiem** [comment]
@wisamidris7 I basically agree with your functional request, but there are some questions, I have commented, please check!

**Review by capdiem — CHANGES_REQUESTED** [review]
There is still some code that has not changed, please make sure the build passes and verify that it is the feature you want.

[feature/treeview-leaf-but-indep-parent](https://github.com/masastack/BlazorComponent/compare/main...feature/treeview-leaf-but-indep-parent) This is the branch where I implemented this feature, and if you don't want to make further changes, you can close this PR.

**Pull request closed without merging**